File: engine/internal/types/ws.go
package types

import "encoding/json"

const (
	MsgTypeGraphUpdate   = "graph_update"
	MsgTypeScoreUpdate   = "score_update"
	MsgTypeChainResult   = "chain_result"
	MsgTypeNarration     = "narration"
	MsgTypeLogEvent      = "log_event"
	MsgTypeThreatIntel   = "threat_intel"
	MsgTypeAgentActivity = "agent_activity"
)

type WSMessage struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

type GraphUpdatePayload struct {
	Nodes []Node `json:"nodes"`
	Edges []Edge `json:"edges"`
}

type ScoreUpdatePayload struct {
	EdgeID string  `json:"edge_id"`
	Score  float64 `json:"score"`
}

type ChainStep struct {
	EventID       string  `json:"event_id"`
	Description   string  `json:"description"`
	Confidence    float64 `json:"confidence"`
	TS            int64   `json:"ts"`
	TechniqueID   string  `json:"technique_id,omitempty"`
	TechniqueName string  `json:"technique_name,omitempty"`
	Tactic        string  `json:"tactic,omitempty"`
}

type ChainResultPayload struct {
	IncidentID    string      `json:"incident_id,omitempty"`
	IncidentLabel string      `json:"incident_label,omitempty"`
	Confidence    float64     `json:"confidence"`
	Steps         []ChainStep `json:"steps"`
	Tactics       []string    `json:"tactics,omitempty"`
}

// NarrationPhase is the AI's prose for one attack phase. Only the id, title,
// and summary come from the model; everything structural about a phase
// (nodes, edges, timestamps, confidence) is derived from the chain by the
// UI so the model cannot hallucinate graph references.
type NarrationPhase struct {
	ID      string `json:"id"`
	Title   string `json:"title"`
	Summary string `json:"summary"`
}

type NarrationPayload struct {
	IncidentID string           `json:"incident_id,omitempty"`
	Text       string           `json:"text"`
	Hypotheses []string         `json:"hypotheses"`
	Actions    []string         `json:"actions"`
	Phases     []NarrationPhase `json:"phases,omitempty"`
}

// LogEventPayload carries the raw Splunk event behind a graph edge so the UI
// can show source log evidence. EventID matches the edge's source_event_id.
type LogEventPayload struct {
	EventID string         `json:"event_id"`
	TS      int64          `json:"ts"`
	Source  string         `json:"source"`
	Raw     map[string]any `json:"raw"`
}

// ThreatIntelPayload is the threat-intel agent's enrichment of a chain. Status
// is "ok" or "error". Techniques are deduplicated by technique ID.
type ThreatIntelPayload struct {
	IncidentID string                 `json:"incident_id,omitempty"`
	Status     string                 `json:"status"`
	Summary    string                 `json:"summary"`
	Techniques []ThreatIntelTechnique `json:"techniques"`
	CVEMatches []CVEMatch             `json:"cve_matches,omitempty"`
}

type ThreatIntelTechnique struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Groups      []string `json:"groups"`
	Software    []string `json:"software"`
	Mitigations []string `json:"mitigations"`
}

type CVEMatch struct {
	ID       string `json:"id"`
	Summary  string `json:"summary"`
	InKEV    bool   `json:"in_kev"`
	Severity string `json:"severity,omitempty"`
}

// AgentActivityPayload is one step in an AI agent's tool-use loop, streamed to
// the UI as it happens so the analyst can watch the narrator and threat-intel
// agent gather and cross-reference evidence in real time. It is purely
// observational: emitting it never changes engine output. Kind is the lifecycle
// stage of the step; Source names the real data source consulted so the feed
// stays honest (e.g. local Splunk vs. an external catalog).
type AgentActivityPayload struct {
	IncidentID string `json:"incident_id,omitempty"`
	// Agent is "narrator" or "intel".
	Agent string `json:"agent"`
	// Seq orders steps within one agent run for a single incident.
	Seq int `json:"seq"`
	// Kind is "thinking" | "tool_call" | "tool_result" | "done" | "error".
	Kind string `json:"kind"`
	// Tool is the tool name for tool_call/tool_result steps, else empty.
	Tool string `json:"tool,omitempty"`
	// Source is the human-readable origin of the data, e.g. "Splunk threat_intel
	// index", "CISA KEV", "NVD", "MITRE ATT&CK".
	Source string `json:"source,omitempty"`
	// Label is a one-line human sentence describing the step.
	Label string `json:"label"`
	// Detail is a short result summary for tool_result steps.
	Detail string `json:"detail,omitempty"`
	// Status is "ok" | "empty" | "error" for tool_result steps.
	Status string `json:"status,omitempty"`
	// TS is the wall-clock unix time the step was emitted (stamped by the pipeline).
	TS int64 `json:"ts,omitempty"`
}

func NewMessage(msgType string, payload any) (WSMessage, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return WSMessage{}, err
	}
	return WSMessage{Type: msgType, Payload: raw}, nil
}

File: ui/src/ws.ts
// WebSocket client for the Causal Reconstruction Engine.
// Plain class (not a React hook) so it can be unit tested without a DOM.

export interface WsNode {
  id: string
  kind: 'Process' | 'Host' | 'User' | 'NetFlow'
  label: string
  attrs: Record<string, unknown>
}

export interface WsEdge {
  id: string
  src: string
  dst: string
  kind: string
  ts: number
  confidence: number
  source_event_id?: string
}

export interface GraphUpdatePayload {
  nodes: WsNode[]
  edges: WsEdge[]
}

export interface ScoreUpdatePayload {
  edge_id: string
  score: number
}

export interface ChainStep {
  event_id: string
  description: string
  confidence: number
  ts: number
  technique_id?: string
  technique_name?: string
  tactic?: string
}

export interface ChainResultPayload {
  incident_id?: string
  incident_label?: string
  confidence: number
  steps: ChainStep[]
  tactics?: string[]
}

export interface ThreatIntelTechnique {
  id: string
  name: string
  groups: string[]
  software: string[]
  mitigations: string[]
}

export interface CVEMatch {
  id: string
  summary: string
  in_kev: boolean
  severity?: string
}

export interface ThreatIntelPayload {
  incident_id?: string
  status: string
  summary: string
  techniques: ThreatIntelTechnique[]
  cve_matches?: CVEMatch[]
}

// Per-phase AI enrichment. The narrator only writes prose; the structural
// truth of a phase (nodes, edges, timestamps, confidence) is derived from the
// chain by the engine and never by the model.
export interface NarrationPhase {
  id: string
  title: string
  summary: string
}

// One prioritized response step. priority is "immediate" | "high" | "medium" |
// "low"; rationale is a one-line justification.
export interface NarrationAction {
  priority: string
  action: string
  rationale: string
}

export interface NarrationPayload {
  incident_id?: string
  text: string
  // Grounded overall severity: "low" | "medium" | "high" | "critical".
  severity?: string
  key_findings?: string[]
  affected_assets?: string[]
  hypotheses: string[]
  actions: NarrationAction[]
  phases?: NarrationPhase[]
}

export interface LogEventPayload {
  event_id: string
  ts: number
  source: string
  raw: Record<string, unknown>
}

// One observational step in an AI agent's tool-use loop, streamed as it happens
// so the UI can show the narrator and threat-intel agent gathering and
// cross-referencing evidence live. Purely informational: it never alters engine
// output. `source` names the real data source consulted, kept honest (local
// Splunk vs. external catalog).
export interface AgentActivityPayload {
  incident_id?: string
  agent: 'narrator' | 'intel'
  seq: number
  kind: 'thinking' | 'tool_call' | 'tool_result' | 'done' | 'error'
  tool?: string
  source?: string
  label: string
  detail?: string
  status?: 'ok' | 'empty' | 'error'
  ts?: number
}

export interface MessageHandlers {
  onGraphUpdate(payload: GraphUpdatePayload): void
  onScoreUpdate(payload: ScoreUpdatePayload): void
  onChainResult(payload: ChainResultPayload): void
  onNarration(payload: NarrationPayload): void
  onLogEvent?(payload: LogEventPayload): void
  onThreatIntel?(payload: ThreatIntelPayload): void
  onAgentActivity?(payload: AgentActivityPayload): void
  onOpen?(): void
  onClose?(): void
}

const INITIAL_RETRY_MS = 1_000
const MAX_RETRY_MS = 30_000

export class EngineSocket {
  private readonly url: string
  private readonly handlers: MessageHandlers
  private readonly wsFactory: (url: string) => WebSocket

  private ws: WebSocket | null = null
  private intentionallyClosed = false
  private retryDelay = INITIAL_RETRY_MS
  private retryTimer: ReturnType<typeof setTimeout> | null = null

  constructor(
    url: string,
    handlers: MessageHandlers,
    wsFactory?: (url: string) => WebSocket,
  ) {
    this.url = url
    this.handlers = handlers
    this.wsFactory = wsFactory ?? ((u) => new WebSocket(u))
  }

  connect(): void {
    this.intentionallyClosed = false
    this.retryDelay = INITIAL_RETRY_MS
    this._open()
  }

  close(): void {
    this.intentionallyClosed = true
    if (this.retryTimer !== null) {
      clearTimeout(this.retryTimer)
      this.retryTimer = null
    }
    this.ws?.close()
    this.ws = null
  }

  private _open(): void {
    const ws = this.wsFactory(this.url)
    this.ws = ws

    ws.onopen = () => {
      this.retryDelay = INITIAL_RETRY_MS
      this.handlers.onOpen?.()
    }

    ws.onmessage = (event: MessageEvent) => {
      let msg: { type: string; payload: unknown }
      try {
        msg = JSON.parse(event.data as string) as { type: string; payload: unknown }
      } catch {
        console.warn('[EngineSocket] unparseable message')
        return
      }
      this._dispatch(msg.type, msg.payload)
    }

    ws.onclose = () => {
      if (this.intentionallyClosed) return
      this.handlers.onClose?.()
      this.retryTimer = setTimeout(() => {
        this.retryDelay = Math.min(this.retryDelay * 2, MAX_RETRY_MS)
        this._open()
      }, this.retryDelay)
    }

    ws.onerror = () => {
      console.error('[EngineSocket] WebSocket error')
    }
  }

  private _dispatch(type: string, payload: unknown): void {
    switch (type) {
      case 'graph_update':
        this.handlers.onGraphUpdate(payload as GraphUpdatePayload)
        break
      case 'score_update':
        this.handlers.onScoreUpdate(payload as ScoreUpdatePayload)
        break
      case 'chain_result':
        this.handlers.onChainResult(payload as ChainResultPayload)
        break
      case 'narration':
        this.handlers.onNarration(payload as NarrationPayload)
        break
      case 'log_event':
        this.handlers.onLogEvent?.(payload as LogEventPayload)
        break
      case 'threat_intel':
        this.handlers.onThreatIntel?.(payload as ThreatIntelPayload)
        break
      case 'agent_activity':
        this.handlers.onAgentActivity?.(payload as AgentActivityPayload)
        break
      default:
        console.warn('[EngineSocket] unknown message type:', type)
    }
  }
}
